### PR TITLE
[MAINT] refactor tests datasets

### LINKED ISSUE #3911: dataset fetch errors without test failures?
Trying to refactor some of the tests I come accross a lot of error messages. I check the latest run in CI and those erros are also there and the test is considered to passs (see example below and attached log).

If this is expected, I find it confusing and it should at least deserve a comment somewhere or a way to reduce the amount of noise in the output.

If this is not expected then, we should probably need to have a way to make those tests fail properly.

<details>
<summary>
Error logs
</summary>
<pre>
2023-08-10T22:54:40.0223742Z Computing word features.
2023-08-10T22:54:40.0224306Z Could not load words from file /tmp/pytest-of-runner/pytest-2/test_fetch_neurovault0/neurovault/collection_993/neurosynth_words_for_image_5432.json; error: Traceback (most recent call last):
2023-08-10T22:54:40.0224885Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 1231, in neurosynth_words_vectorized
2023-08-10T22:54:40.0225318Z     info = json.loads(word_file.read().decode("utf-8"))
2023-08-10T22:54:40.0225578Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0225913Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/__init__.py", line 346, in loads
2023-08-10T22:54:40.0226234Z     return _default_decoder.decode(s)
2023-08-10T22:54:40.0226466Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0226931Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 337, in decode
2023-08-10T22:54:40.0227245Z     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
2023-08-10T22:54:40.0227493Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0227829Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 355, in raw_decode
2023-08-10T22:54:40.0228231Z     raise JSONDecodeError("Expecting value", s, err.value) from None
2023-08-10T22:54:40.0228573Z json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
2023-08-10T22:54:40.0228776Z 
2023-08-10T22:54:40.0229278Z Could not load words from file /tmp/pytest-of-runner/pytest-2/test_fetch_neurovault0/neurovault/collection_993/neurosynth_words_for_image_7492.json; error: Traceback (most recent call last):
2023-08-10T22:54:40.0229844Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 1231, in neurosynth_words_vectorized
2023-08-10T22:54:40.0230276Z     info = json.loads(word_file.read().decode("utf-8"))
2023-08-10T22:54:40.0230533Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0230855Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/__init__.py", line 346, in loads
2023-08-10T22:54:40.0231164Z     return _default_decoder.decode(s)
2023-08-10T22:54:40.0231380Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0231697Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 337, in decode
2023-08-10T22:54:40.0232009Z     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
2023-08-10T22:54:40.0232245Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0232567Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 355, in raw_decode
2023-08-10T22:54:40.0233085Z     raise JSONDecodeError("Expecting value", s, err.value) from None
2023-08-10T22:54:40.0233424Z json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

...

2023-08-10T22:54:40.0273769Z 
2023-08-10T22:54:40.0274219Z Could not load words from file /tmp/pytest-of-runner/pytest-2/test_fetch_neurovault0/neurovault/collection_859/neurosynth_words_for_image_4351.json; error: Traceback (most recent call last):
2023-08-10T22:54:40.0274772Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 1231, in neurosynth_words_vectorized
2023-08-10T22:54:40.0275189Z     info = json.loads(word_file.read().decode("utf-8"))
2023-08-10T22:54:40.0275430Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0275753Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/__init__.py", line 346, in loads
2023-08-10T22:54:40.0276059Z     return _default_decoder.decode(s)
2023-08-10T22:54:40.0276271Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0276592Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 337, in decode
2023-08-10T22:54:40.0276913Z     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
2023-08-10T22:54:40.0277167Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0277479Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 355, in raw_decode
2023-08-10T22:54:40.0277872Z     raise JSONDecodeError("Expecting value", s, err.value) from None
2023-08-10T22:54:40.0278214Z json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
2023-08-10T22:54:40.0278413Z 
2023-08-10T22:54:40.0278861Z Could not load words from file /tmp/pytest-of-runner/pytest-2/test_fetch_neurovault0/neurovault/collection_859/neurosynth_words_for_image_9351.json; error: Traceback (most recent call last):
2023-08-10T22:54:40.0279415Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 1231, in neurosynth_words_vectorized
2023-08-10T22:54:40.0279815Z     info = json.loads(word_file.read().decode("utf-8"))
2023-08-10T22:54:40.0280171Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0280506Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/__init__.py", line 346, in loads
2023-08-10T22:54:40.0280821Z     return _default_decoder.decode(s)
2023-08-10T22:54:40.0281033Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0281357Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 337, in decode
2023-08-10T22:54:40.0281685Z     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
2023-08-10T22:54:40.0281933Z                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-08-10T22:54:40.0282251Z   File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/json/decoder.py", line 355, in raw_decode
2023-08-10T22:54:40.0282637Z     raise JSONDecodeError("Expecting value", s, err.value) from None
2023-08-10T22:54:40.0283035Z json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
2023-08-10T22:54:40.0283235Z 
2023-08-10T22:54:40.0283243Z 
2023-08-10T22:54:40.0283548Z Dataset created in /tmp/pytest-of-runner/pytest-2/temp_nilearn_home375/nilearn_shared_data/neurovault
2023-08-10T22:54:40.0283774Z 
2023-08-10T22:54:40.0283871Z Reading local neurovault data.
2023-08-10T22:54:40.0284093Z No image found on local disk.
2023-08-10T22:54:40.0284442Z fetch_neurovault: using default value of 100 for max_images. Set max_images to another value or None if you want more images.
2023-08-10T22:54:40.0284761Z Reading local neurovault data.
2023-08-10T22:54:40.0284983Z Already fetched 1 image
2023-08-10T22:54:40.0285190Z Already fetched 2 images
2023-08-10T22:54:40.0285392Z Already fetched 3 images
2023-08-10T22:54:40.0285592Z Already fetched 4 images
2023-08-10T22:54:40.0285793Z Already fetched 5 images
2023-08-10T22:54:40.0285989Z Already fetched 6 images
2023-08-10T22:54:40.0286172Z Already fetched 7 images
2023-08-10T22:54:40.0286368Z Already fetched 8 images
2023-08-10T22:54:40.0286563Z Already fetched 9 images
2023-08-10T22:54:40.0286757Z Already fetched 10 images
2023-08-10T22:54:40.0286964Z Already fetched 11 images
2023-08-10T22:54:40.0287300Z 11 images found on local disk.
2023-08-10T22:54:40.0287496Z PASSED
2023-08-10T22:54:40.0364054Z Traceback (most recent call last):
2023-08-10T22:54:40.0364567Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 996, in _get_batch
2023-08-10T22:54:40.0370665Z     resp.raise_for_status()
2023-08-10T22:54:40.0375992Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/_testing.py", line 129, in raise_for_status
2023-08-10T22:54:40.0376415Z     raise HTTPError(f"{self.status_code} Error for url: {self.url}")
2023-08-10T22:54:40.0382204Z requests.exceptions.HTTPError: 500 Error for url: https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.0382588Z Traceback (most recent call last):
2023-08-10T22:54:40.0382936Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 996, in _get_batch

...

2023-08-10T22:54:40.0961854Z     raise HTTPError(f"{self.status_code} Error for url: {self.url}")
2023-08-10T22:54:40.0962231Z requests.exceptions.HTTPError: 500 Error for url: https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.0962811Z nilearn/datasets/tests/test_neurovault.py::test_fetch_neurovault_errors fetch_neurovault: using default value of 100 for max_images. Set max_images to another value or None if you want more images.
2023-08-10T22:54:40.0963150Z 
2023-08-10T22:54:40.0963631Z Dataset created in /tmp/pytest-of-runner/pytest-2/temp_nilearn_home376/nilearn_shared_data/neurovault
2023-08-10T22:54:40.0963870Z 
2023-08-10T22:54:40.0963971Z Reading local neurovault data.
2023-08-10T22:54:40.0964189Z No image found on local disk.
2023-08-10T22:54:40.0964416Z Reading server neurovault data.
2023-08-10T22:54:40.0964744Z Scroll collections: getting new batch: https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.0965140Z Could not get batch from https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.0965533Z Scroll collections: getting new batch: https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.0965921Z Could not get batch from https://neurovault.org/api/collections/?limit=100&offset=0

...

2023-08-10T22:54:40.0998124Z Traceback (most recent call last):
2023-08-10T22:54:40.0998491Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 996, in _get_batch
2023-08-10T22:54:40.0998797Z     resp.raise_for_status()
2023-08-10T22:54:40.0999140Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/_testing.py", line 129, in raise_for_status

...

2023-08-10T22:54:40.1059621Z     raise HTTPError(f"{self.status_code} Error for url: {self.url}")
2023-08-10T22:54:40.1059840Z requests.exceptions.HTTPError: 500 Error for url: https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.1059940Z Traceback (most recent call last):
2023-08-10T22:54:40.1060142Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/neurovault.py", line 996, in _get_batch
2023-08-10T22:54:40.1060235Z     resp.raise_for_status()
2023-08-10T22:54:40.1060444Z   File "/home/runner/work/nilearn/nilearn/nilearn/datasets/_testing.py", line 129, in raise_for_status
2023-08-10T22:54:40.1060629Z     raise HTTPError(f"{self.status_code} Error for url: {self.url}")
2023-08-10T22:54:40.1060840Z requests.exceptions.HTTPError: 500 Error for url: https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.1061056Z Scroll collections: getting new batch: https://neurovault.org/api/collections/?limit=100&offset=0
2023-08-10T22:54:40.1061238Z Could not get batch from https://neurovault.org/api/collections/?limit=100&offset=0

...

2023-08-10T22:54:40.1108125Z PASSED
2023-08-10T22:54:40.2207873Z nilearn/datasets/tests/test_neurovault.py::test_fetch_neurovault_ids 
2023-08-10T22:54:40.2208406Z Dataset created in /tmp/pytest-of-runner/pytest-2/temp_nilearn_home377/nilearn_shared_data/neurovault
2023-08-10T22:54:40.2208418Z 
2023-08-10T22:54:40.2208425Z 
2023-08-10T22:54:40.2208704Z Dataset created in /tmp/pytest-of-runner/pytest-2/test_fetch_neurovault_ids0/neurovault
2023-08-10T22:54:40.2208715Z 
2023-08-10T22:54:40.2208818Z Reading local neurovault data.
2023-08-10T22:54:40.2208932Z No image found on local disk.
2023-08-10T22:54:40.2209022Z Reading server neurovault data.
</pre>
</details>

[8_Run tests.txt](https://github.com/nilearn/nilearn/files/12337967/8_Run.tests.txt)

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Relates to #3570
- Closes #3911 

Changes proposed in this pull request:

- refactor complex tests in nilearn/datasets/tests

### CODE REVIEW

File: nilearn/datasets/tests/test_atlas.py [hunk 1]
```diff
@@ -624,41 +552,44 @@ def test_fetch_atlas_basc_multiscale_2015(tmp_path, request_mocker):
...
 
     name_asym = "template_cambridge_basc_multiscale_nii_asym"
-    basenames_asym = [
-        f"template_cambridge_basc_multiscale_asym_{key}.nii.gz" for key in keys
-    ]
-    for key, basename_asym in zip(keys, basenames_asym):
-        assert data_asym[key] == str(
-            tmp_path / dataset_name / name_asym / basename_asym
-        )
+    basename_sym = f"template_cambridge_basc_multiscale_asym_{key}.nii.gz"
```
Comment (bthirion) [inline nilearn/datasets/tests/test_atlas.py:579]:
```suggestion
    basename_asym = f"template_cambridge_basc_multiscale_asym_{key}.nii.gz"
```

File: nilearn/datasets/tests/test_atlas.py [hunk 2]
```diff
@@ -624,41 +552,44 @@ def test_fetch_atlas_basc_multiscale_2015(tmp_path, request_mocker):
...
+
+    assert data_asym[key] == str(
+        tmp_path / dataset_name / name_asym / basename_sym
```
Comment (bthirion) [inline nilearn/datasets/tests/test_atlas.py:582]:
```suggestion
        tmp_path / dataset_name / name_asym / basename_asym
```

File: nilearn/datasets/tests/test_func.py [hunk 3]
```diff
@@ -372,7 +375,7 @@ def test_fetch_abide_pcp(tmp_path, request_mocker, quality_checked):
     )
 
 
-def test__load_mixed_gambles(request_mocker):
```
Comment (bthirion) [inline nilearn/datasets/tests/test_func.py:375]: How do you get rid of request_mocker ?

**Review by bthirion — COMMENTED** [review]
Thx for tackling this !

File: nilearn/datasets/tests/test_func.py (hunk 3 above)
Comment (Remi-Gau) [reply nilearn/datasets/tests/test_func.py:375]: it is set to autouse so it gets called in any case before any test

https://github.com/nilearn/nilearn/blob/0b49cc0cd99ca5bfd8780e5429c5d744dc21db81/nilearn/datasets/_testing.py#L69

**Comment by Remi-Gau** [comment]
- [ ] address #3911

**Comment by Remi-Gau** [comment]
Sorry for the large number of line changes: a lot of this is due to adding new lines to separate the set up, test, teardown of each tests. I find that it improves test readability.

File: nilearn/datasets/tests/test_neurovault.py [hunk 4]
```diff
@@ -712,13 +776,20 @@ def test_fetch_neurovault(tmp_path):
     os.chmod(os.path.join(tmp_path, "neurovault"), stat.S_IREAD | stat.S_IEXEC)
     if os.access(os.path.join(tmp_path, "neurovault"), os.W_OK):
         return
+
     with pytest.warns(UserWarning):
         neurovault.fetch_neurovault(data_dir=tmp_path)
 
 
 def test_fetch_neurovault_errors(request_mocker):
+    """Test that errors are raised when the server returns an error code.
```
Comment (ymzayek) [inline nilearn/datasets/tests/test_neurovault.py:785]: Maybe more accurate

```suggestion
    """Test that errors are logged when the server returns an error code.
```

Can we instead capture the error with pytest's [capsys](https://docs.pytest.org/en/7.4.x/reference/reference.html#capsys) fixture and match it to a pattern in the error? Something like:
```python
def test_fetch_neurovault_errors(capsys):
    request_mocker.url_mapping["*"] = 500
    data = neurovault.fetch_neurovault()
    captured = capsys.readouterr()
    match = re.search("500 Error", captured.err)
    assert match is not None
    assert len(data.images) == 0
```

Comment (Remi-Gau) [reply nilearn/datasets/tests/test_neurovault.py:785]: Uuuh. Nice idea. Me like it.

File: nilearn/datasets/tests/test_atlas.py [hunk 5]
```diff
@@ -16,76 +15,12 @@
...
 from nilearn._utils import data_gen
 from nilearn._utils.testing import serialize_niimg
-from nilearn.datasets import atlas, utils
+from nilearn.datasets import atlas
 from nilearn.datasets._testing import dict_to_archive
+from nilearn.datasets.utils import _fetch_files
 from nilearn.image import get_data
 
 
-def test_get_dataset_dir(tmp_path):
```
Comment (bthirion) [inline nilearn/datasets/tests/test_atlas.py:24]: So, this test is no longer necessary ?

File: nilearn/datasets/tests/test_struct.py [hunk 6]
```diff
@@ -2,83 +2,20 @@
...
 import pandas as pd
 import pytest
+from nibabel import Nifti1Image
 
-from nilearn.datasets import struct, utils
+from nilearn.datasets import struct
 from nilearn.datasets._testing import dict_to_archive, list_to_archive
 
 
-def test_get_dataset_dir(tmp_path):
```
Comment (bthirion) [inline nilearn/datasets/tests/test_struct.py:18]: Same thing here.

**Review by bthirion — COMMENTED** [review]
I think that this is mostly good. Just making sure I understand the diff...

File: nilearn/datasets/tests/test_atlas.py (hunk 5 above)
Comment (Remi-Gau) [reply nilearn/datasets/tests/test_atlas.py:24]: As you guessed this test was duplicated: see [here](https://github.com/nilearn/nilearn/issues/3570#issuecomment-1677562849)

File: nilearn/datasets/tests/test_struct.py (hunk 6 above)
Comment (Remi-Gau) [reply nilearn/datasets/tests/test_struct.py:18]: As you guessed this test was duplicated: see [here](https://github.com/nilearn/nilearn/issues/3570#issuecomment-1677562849)

File: nilearn/datasets/tests/test_neurovault.py (hunk 4 above)
Comment (Remi-Gau) [reply nilearn/datasets/tests/test_neurovault.py]: done

**Review by bthirion — APPROVED** [review]
LGTM, thx !

**Pull request merged**